cranelift/docs/isle-integration.md

```diff
@@ -0,0 +1,188 @@
+`cranelift/codegen/src/isa/<arch>/inst.isle` and general Rust code (either in
+`cranelift/codegen/src/isa/<arch>/lower/isle.rs` or elsewhere).
+
+When an instruction modifies a register, both reading from it and writing to it,
+we should build an SSA view of that instruction that gets legalized via "move
+mitosis" by splitting a move out from the register.
+
+For example, on x86 the `add` instruction reads and writes its first operand:
+
+    add a, b    ==    a = a + b
+
+So we present an SSA facade where `add` operates on three registers, instead of
+two, and defines one of them, while reading the other two and leaving them
+unmodified:
+
+    add a, b, c    ==    a = b + c
+
+Then, as an implementation detail of the facade, we emit moves as necessary:
+
+    add a, b, c    ==>    mov a, b; add b, c
+
+We call the process of emitting these moves "move mitosis". For ISAs with
+ubiquitous use of modified registers and instructions in two-operand form, like
+x86, we implement move mitosis with methods on the ISA's `MachInst`. For other
+ISAs that are RISCier and where modified registers are pretty rare, such as
+aarch64, we implement the handful of move mitosis special cases at the
+`inst.isle` layer. Either way, the important thing is that the lowering rules
+remain pure.
+
+Fnally, note that these moves are generally cleaned up by the register
+allocator's move coalescing, and move mitosis will eventually go away completely
+once we switch over to `regalloc2`, which takes instructions in SSA form
+directly as input.
+
+Instructions that implicitly operate on specific registers, or which require
+that certain operands be in certain registers, are handled similarly: the
+lowering rules use a pure paradigm that ignores these constraints and has
+instructions that explicitly take implicit operands, and we ensure the
+constraints are fulfilled a layer below the lowering rules (in `inst.isle` or in
+Rust glue code).
+
+## When are lowering rules allowed to have side effects?
+
+Extractors (the matchers that appear on the left-hand sides of `rule`s) should
+**never** have side effects. When evaluating a rule's extractors, we haven't yet
+committed to evaluating that rule's right-hand side. If the extractors performed
+side effects, we could get deeply confusing action-at-a-distance bugs where
+rules we never fully match pull the rug out from under our feet.
+
+Anytime you are tempted to perform side effects in an extractor, you should
+instead just package up the things you would need in order to perform that side
+effect, and then have a separate constructor that takes that package and
+performs the side effect it describes. The constructor can only be called inside
+a rule's right-hand side, which is only evaluated after we've committed to this
+rule, which avoids the action-at-a-distance bugs described earlier.
+
+For example, loads have a side effect in CLIF: they might trap. Therefore, even
```

Comment:
This paragraph has good detail but is a little dense; a snippet or two showing how `sinkable_load` and `sink_load` are used in a rule might help?